File: 2022/02/01/validParentheses.swift
// https://leetcode.com/problems/valid-parentheses/
// Author: Andrew Jarombek
// Date: 2/1/2022

func isValid(_ s: String) -> Bool {
    var items = [String]()
    let pairs = [
        "(": ")",
        "[": "]",
        "{": "}"
    ]

    for c in s {
        if let value = pairs[String(c)] {
            items.append(value)
        } else if items.count > 0 && String(c) == items[items.count - 1] {
            items.removeLast()
        } else {
            return false
        }
    }

    return items.count == 0
}

File: 2022/02/12/cpp/validParentheses.cpp
// https://leetcode.com/problems/valid-parentheses/
// Author: Andrew Jarombek
// Date: 2/12/2022

#include <string>
#include <stack>
#include <map>
#include <cassert>

using namespace std;

bool isValid(string s) {
    stack<char> closingParens;
    map<char, char> parens = {
            {'(', ')'},
            {'[', ']'},
            {'{', '}'}
    };

    for (int i = 0; i < s.length(); i++) {
        char c = s[i];

        if (parens.find(c) != parens.end()) {
            closingParens.push(parens[c]);
        } else {
            if (closingParens.size() == 0 || c != closingParens.top()) {
                return false;
            }

            closingParens.pop();
        }
    }

    return closingParens.size() == 0;
}

int main() {
    assert(isValid("()"));
    assert(isValid("()[]{}"));
    assert(!isValid("[}"));
}
